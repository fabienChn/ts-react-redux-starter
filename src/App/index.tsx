import { hot } from 'react-hot-loader/root';
import React, { useState, useEffect } from 'react';
import { Switch } from 'react-router-dom';
import { Route } from 'react-router';
import styled from 'styled-components';

import ConnectApp, { ConnectAppProps } from 'connectors/App';
import { StateHook } from 'types';
import { ITheme } from 'theme';

const Container = styled.main<{ theme: ITheme }>`
  border: 1px solid ${({ theme }: { theme: ITheme }): string => theme.colors.red};
`;

const App = (props: ConnectAppProps): JSX.Element => {
  const [title, setTitle]: StateHook<string> = useState('My App');

  useEffect((): void => {
    // tslint:disable-next-line:no-console
    console.log(props);

    setTimeout((): void => {
      setTitle('The title changed haha');
    }, 2000);
  }, []);

  return (
    <Container className="App">
      <>
        <div>
          <h1>{title}</h1>
          <span>hehe</span>
          <Switch>
            <Route
              exact
              path="/settings"
              render={(): JSX.Element => <div />}
            />

            <Route
              exact
              path="/scooters/:id?"
              render={(): JSX.Element => <div />}
            />
          </Switch>
        </div>
      </>
    </Container>
  );
};

// export hot reloading enabled if it is development
const exportedApp = process.env.NODE_ENV === 'development' ? hot(App) : App;
export default exportedApp;

const connectedExport =
  process.env.NODE_ENV === 'development'
    ? hot(ConnectApp(App))
    : ConnectApp(exportedApp);
export const ConnectedApp = connectedExport;
