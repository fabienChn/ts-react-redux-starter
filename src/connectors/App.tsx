import { ComponentClass } from 'react';
import { bindActionCreators, Dispatch } from 'redux';
import { connect } from 'react-redux';
import * as actions from 'store/actions';
import { IStoreState, IEventsState, IDispatchProps } from 'store/types';

export interface IStateProps {
  events?: IEventsState;
}

export type ConnectAppProps = IDispatchProps & IStateProps;

export default <T extends {}>(
  Component: any,
): ComponentClass<T & ConnectAppProps> => {
  const mapStateToProps = (state: IStoreState): IStateProps => {
    return {
      events: state.events,
    };
  };

  const mapDispatchToProps = (dispatch: Dispatch): IDispatchProps => {
    return {
      actions: bindActionCreators(actions, dispatch),
    };
  };

  // @ts-ignore
  return connect<T & ConnectAppProps>(
    // @ts-ignore
    mapStateToProps,
    mapDispatchToProps,
  )(Component);
};
