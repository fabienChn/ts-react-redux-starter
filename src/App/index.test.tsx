import { ShallowWrapper } from 'enzyme';
import {
  createShallowWrapper,
  expectWrappersTreeToMatchSnapshot,
} from 'utils/testHelpers';
import App from './index';

const createSubject = (): ShallowWrapper => createShallowWrapper(App, {});

it('renders the expected snapshot', (): void => {
  expectWrappersTreeToMatchSnapshot(createSubject());
});

it('should change the title after 2 seconds', (): void => {
  const wrapper = createSubject();

  expect(wrapper.text()).toContain('My App');

  setTimeout((): void => {
    expect(wrapper.text()).toContain('The title changed haha');
  }, 2000);
});