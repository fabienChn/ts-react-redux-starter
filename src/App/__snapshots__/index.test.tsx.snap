// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the expected snapshot 1`] = `
<styled.main
  className="App"
>
  <div>
    <h1>
      My App
    </h1>
    <span>
      hehe
    </span>
    <Switch>
      <Route
        exact={true}
        path="/settings"
        render={[Function]}
      />
      <Route
        exact={true}
        path="/scooters/:id?"
        render={[Function]}
      />
    </Switch>
  </div>
</styled.main>
`;
